src/sdf/include/sdf.h

```diff
@@ -0,0 +1,474 @@
+         (SpreadBits3(index.y) << 2) | (SpreadBits3(index.z) << 3);
+}
+
+__host__ __device__ glm::ivec4 DecodeMorton(Uint64 code) {
+  glm::ivec4 index;
+  index.x = SqeezeBits3(code >> 1);
+  index.y = SqeezeBits3(code >> 2);
+  index.z = SqeezeBits3(code >> 3);
+  index.w = code & 0x1u;
+  return index;
+}
+
+struct GridVert {
+  Uint64 key = kOpen;
+  float distance = NAN;
+  int edgeIndex = -1;
+  int edgeVerts[7] = {-1, -1, -1, -1, -1, -1, -1};
+
+  __host__ __device__ int Inside() const { return distance > 0 ? 1 : -1; }
+};
+
+class HashTableD {
+ public:
+  HashTableD(VecDH<GridVert>& alloc, VecDH<uint32_t>& used, uint32_t step = 127)
+      : step_{step}, table_{alloc}, used_{used} {}
+
+  __host__ __device__ int Size() const { return table_.size(); }
+
+  __host__ __device__ bool Insert(const GridVert& vert) {
+    uint32_t idx = vert.key & (Size() - 1);
+    while (1) {
+      Uint64& key = table_[idx].key;
+      const Uint64 found = AtomicCAS(key, kOpen, vert.key);
+      if (found == kOpen) {
+        if (AtomicAdd(used_[0], 0x1u) * 2 > Size()) {
+          return true;
+        }
+        table_[idx] = vert;
+        return false;
+      }
+      if (found == vert.key) return false;
+      idx = (idx + step_) & (Size() - 1);
+    }
+  }
+
+  __host__ __device__ GridVert operator[](Uint64 key) const {
+    uint32_t idx = key & (Size() - 1);
+    while (1) {
+      const GridVert found = table_[idx];
+      if (found.key == key || found.key == kOpen) return found;
+      idx = (idx + step_) & (Size() - 1);
+    }
+  }
+
+  __host__ __device__ GridVert At(int idx) const { return table_[idx]; }
+
+ private:
+  const uint32_t step_;
+  VecD<GridVert> table_;
+  VecD<uint32_t> used_;
+};
+
+class HashTable {
+ public:
+  HashTable(uint32_t size, uint32_t step = 127)
+      : alloc_{1 << (int)ceil(log2(size)), {}}, table_{alloc_, used_, step} {}
+
+  HashTableD D() { return table_; }
+
+  int Entries() const { return used_[0]; }
+
+  int Size() const { return table_.Size(); }
+
+  float FilledFraction() const {
+    return static_cast<float>(used_[0]) / table_.Size();
+  }
+
+ private:
+  VecDH<GridVert> alloc_;
+  VecDH<uint32_t> used_ = VecDH<uint32_t>(1, 0);
+  HashTableD table_;
+};
+
+template <typename Func>
+struct ComputeVerts {
+  glm::vec3* vertPos;
+  int* vertIndex;
+  HashTableD gridVerts;
+  const Func sdf;
+  const glm::vec3 origin;
+  const glm::ivec3 gridSize;
+  const glm::vec3 spacing;
+  const float level;
+
+  inline __host__ __device__ glm::vec3 Position(glm::ivec4 gridIndex) const {
+    return origin + spacing * (-0.5f + glm::vec3(gridIndex) +
+                               (gridIndex.w == 1 ? 0.5f : 0.0f) * glm::vec3(1));
+  }
+
+  inline __host__ __device__ float BoundedSDF(glm::ivec4 gridIndex) const {
+    const float d = sdf(Position(gridIndex)) - level;
+
+    const glm::ivec3 xyz(gridIndex);
+    if (glm::any(glm::equal(xyz, glm::ivec3(0))) ||
```

Comment:
Maybe you can assign a boolean variable for these conditions to make it more readable? The condition is quite complicated here.